task.go

```diff
@@ -202,39 +209,103 @@ func Build(buildkitd *Buildkitd, outputsDir string, req Request) (Response, erro
 		}
 	}
 
+	if req.Config.OutputOCI {
+		err = loadOciImages(imagePaths, req)
+		if err != nil {
+			return Response{}, err
+		}
+	} else {
+		err = loadImages(imagePaths, req)
+		if err != nil {
+			return Response{}, err
+		}
+	}
+
+	return res, nil
+}
+
+func loadImages(imagePaths []string, req Request) error {
 	for _, imagePath := range imagePaths {
 		image, err := tarball.ImageFromPath(imagePath, nil)
 		if err != nil {
-			return Response{}, errors.Wrap(err, "open oci image")
+			return errors.Wrap(err, "open oci image")
 		}
 
 		outputDir := filepath.Dir(imagePath)
 
 		err = writeDigest(outputDir, image)
 		if err != nil {
-			return Response{}, err
+			return err
 		}
 
 		if req.Config.UnpackRootfs {
-			err = unpackRootfs(outputDir, image, cfg)
+			err = unpackRootfs(outputDir, image, req.Config)
 			if err != nil {
-				return Response{}, errors.Wrap(err, "unpack rootfs")
+				return errors.Wrap(err, "unpack rootfs")
 			}
 		}
 	}
+	return nil
+}
 
-	return res, nil
+func loadOciImages(imagePaths []string, req Request) error {
+	for _, imagePath := range imagePaths {
+		_, err := os.Stat(imagePath)
+		if err != nil {
+			return errors.Wrapf(err, "image path %s not valid", imagePath)
+		}
+
+		// go-containerregistry does not currently have support for loading a OCI formated
+		// image from a tarball, so we uncompress it before doing anything.
+		logrus.Info("uncompressing OCI image tar")
+		targetDir := filepath.Dir(imagePath)
+		imageDir := filepath.Join(targetDir, "image")
+		err = os.MkdirAll(imageDir, 0700)
+		if err != nil {
+			return errors.Wrapf(err, "unable to create image dir %s", imageDir)
+		}
+		run(os.Stdout, "tar", "-xvf", imagePath, "-C", imageDir)
+
+		l, err := layout.ImageIndexFromPath(imageDir)
+		if err != nil {
+			return errors.Wrapf(err, "failed to load %s as OCI layout", imagePath)
+		}
+
+		m, err := l.IndexManifest()
+		if err != nil {
+			return errors.Wrap(err, "error getting index manifest")
+		}
+
+		var image partial.WithRawManifest
+		desc := m.Manifests[0]
+		if desc.MediaType.IsImage() {
+			image, err = l.Image(desc.Digest)
+		} else if desc.MediaType.IsIndex() {
+			image, err = l.ImageIndex(desc.Digest)
+		} else {
+			return fmt.Errorf("layout contains non-image (mediaType: %q)", desc.MediaType)
+		}
+
+		outputDir := filepath.Dir(imagePath)
+
+		err = writeDigest(outputDir, image)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
-func writeDigest(dest string, image v1.Image) error {
+func writeDigest(dest string, image partial.WithRawManifest) error {
```

Comment:
Whats the difference between v1.Image and partial.WithRawManifest? More general what is the usage of partial? Thx

Reply:
The definition of v1.Image and v1.ImageIndex is incompatible so we use partial.WithRawManifest which has a compatible subset between the two so we can pass either  v1.Image or v1.ImageIndex  here when writing the digest. 

As for the more general usage of partial their docs does a way better job at explaining it than I could: https://github.com/google/go-containerregistry/tree/main/pkg/v1/partial

Reply:
I was actually able to simplify things and remove the utilization of partial but simply passing the digest directly into writeDigest between the two load functions.

Reply:
I was actually able to simplify things and removing the usage of partial by extracting the digest before sending it to be written.